aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp

```diff
@@ -0,0 +1,138 @@
+          lrintf((input_row[col] - minimum_element) * inverse_scale);
+    }
+  }
+  return output;
+}
+
+Tensor qembeddingbag_4bit_prepack(const Tensor& weight) {
+  int64_t embedding_rows = weight.size(0);
+  int64_t embedding_cols = weight.size(1);
+
+  Tensor weight_contig = weight.contiguous(weight.suggest_memory_format());
+
+  const auto weight_data = weight.data_ptr<float>();
+  constexpr int BIT_RATE = 4;
+  constexpr int NUM_ELEM_PER_BYTE = 8 / BIT_RATE;
+  TORCH_CHECK(
+      weight_contig.size(weight.dim() - 1) % NUM_ELEM_PER_BYTE == 0,
+      "FloatToFused4BitRowwiseQuantizedOp only works for the number of "
+      "columns a multiple of 2");
+
+  // The "fused" representation stores the scale and bias with the
+  // row-wise quantized data in one tensor.
+  // Since we represent the scale and bias in 16-bit float, we'll use the
+  // last 4 bytes of each row for scale (2 bytes) and bias (2 bytes).
+  // | ... quantized data ... | scale | bias |
+  // |    number_of_columns   |  2B   |  2B  |
+  std::vector<int64_t> output_shape = {
+      embedding_rows,
+      static_cast<std::int64_t>(
+          (embedding_cols + NUM_ELEM_PER_BYTE - 1) / NUM_ELEM_PER_BYTE +
+          2 * sizeof(at::Half))};
+  auto output = at::empty(
+      output_shape,
+      weight_contig.options().dtype(at::kByte),
+      weight_contig.suggest_memory_format());
+  auto* output_data = output.data_ptr<uint8_t>();
+  const auto output_columns = output.size(output.dim() - 1);
+
+  for (int row = 0; row < embedding_rows; ++row) {
+    const float* input_row = weight_data + row * embedding_cols;
+    std::uint8_t* output_row = output_data + row * output_columns;
+
+    at::Half* output_row_scale_zp = reinterpret_cast<at::Half*>(
+        output_row +
+        (embedding_cols + NUM_ELEM_PER_BYTE - 1) / NUM_ELEM_PER_BYTE);
+
+    float Xmin = *std::min_element(input_row, input_row + embedding_cols);
+    float Xmax = *std::max_element(input_row, input_row + embedding_cols);
+
+    Xmin = static_cast<at::Half>(Xmin);
+    const float range = Xmax - Xmin;
+
+    // Set scale to 1.0f for the corner case of Xmax == Xmin .
+    // Any non-zero scale would work because during quantization
+    // (X - Xmin) / scale will be 0 for all X unless scale is 0.
+    at::Half scale = range == 0 ? 1.0f : range / ((1 << BIT_RATE) - 1);
+    float inverse_scale = scale == 0 ? 1.0f : 1.0f / scale;
+    if (scale == 0 || std::isinf(inverse_scale)) {
+      // Corner case handling when Xmax == Xmin
+      // Any scale would work because X - Xmin will be 0 for all X
+      scale = 1.0f;
+      inverse_scale = 1.0f;
+    }
+    output_row_scale_zp[0] = scale;
+    output_row_scale_zp[1] = Xmin;
+
+    for (int col = 0; col < embedding_cols; ++col) {
+      float X = input_row[col];
+      std::uint8_t quantized = std::max(
+          0,
+          std::min<int>(
+              lrintf((X - Xmin) * inverse_scale), (1 << BIT_RATE) - 1));
+      if (col % NUM_ELEM_PER_BYTE == 0) {
```

Comment:
maybe we can add a quick comment on the data format? i.e. index 0 is packed in the low bits and index 1 in the high bits, etc